File: lcof/面试题07. 重建二叉树/Solution.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func buildTree(preorder []int, inorder []int) *TreeNode {
	d := map[int]int{}
	for i, v := range inorder {
		d[v] = i
	}
	var dfs func(i, j, n int) *TreeNode
	dfs = func(i, j, n int) *TreeNode {
		if n < 1 {
			return nil
		}
		k := d[preorder[i]]
		l := k - j
		root := &TreeNode{Val: preorder[i]}
		root.Left = dfs(i+1, j, l)
		root.Right = dfs(i+1+l, k+1, n-l-1)
		return root
	}
	return dfs(0, 0, len(inorder))
}

File: lcof/面试题07. 重建二叉树/Solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function (preorder, inorder) {
    const d = new Map();
    const n = inorder.length;
    for (let i = 0; i < n; ++i) {
        d.set(inorder[i], i);
    }
    const dfs = (i, j, n) => {
        if (n < 1) {
            return null;
        }
        const k = d.get(preorder[i]);
        const l = k - j;
        const root = new TreeNode(preorder[i]);
        root.left = dfs(i + 1, j, l);
        root.right = dfs(i + 1 + l, k + 1, n - l - 1);
        return root;
    };
    return dfs(0, 0, n);
};
